Separate debug configuration providers from debug adapter contributors

alvsan09 opened the pull request:
#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

Fixes: #9978 

The implementation of the vscode API for the resolution of initial
debug configurations as well as the resolution of specific
configurations was relying on registered debug adapter
contributors rather than debug configuration providers,
this is causing issues e.g. #9978.

This implementation brings access to the debug configuration
providers all the way to the browser debug/plugin-debug-service,
where the providers can be filtered and then call its corresponding
API indirectly towards the extension host.

The former API is kept and marked deprecated.

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
Based on: https://github.com/eclipse-theia/theia/pull/5645
Use the test extension in the `tar` file as follows:
https://github.com/alvsan09/debug-activation-events/raw/master/plugins.tar

NOTE: It is recommended to install only the required test extension(s), otherwise not all notifications may appear (must likely due to a hard coded limit).

- test `onDebug`: install [on-debug-0.0.1.vsix](https://github.com/alvsan09/debug-activation-events/blob/master/on-debug-0.0.1.vsix) and check that `onDebug` notification appears on any debug activation event
- test `onDebugInitialConfigurations`:
  - install [on-debug-initial-configurations-0.0.1.vsix](https://github.com/alvsan09/debug-activation-events/blob/master/on-debug-initial-configurations-0.0.1.vsix)
  - delete launch.json file
  - try to add a new configuration with markdown file opened 
  - you should see `onDebugInitialConfigurations` and `provideDebugConfigurations` notifications plus mock debug configuration should be generated dynamically
- test `onDebugResolve`:
  - install [on-debug-resolve-0.0.1.vsix](https://github.com/alvsan09/debug-activation-events/blob/master/on-debug-resolve-0.0.1.vsix)
  - try to run any launch configuration
  - you should see `onDebugResolve` and `resolveDebugConfiguration(*)[0]` notifications
- test `onDebugResolve:<debug type>`:
  - install [on-debug-resolve-node-0.0.1.vsix](https://github.com/alvsan09/debug-activation-events/blob/master/on-debug-resolve-node-0.0.1.vsix)
  - try to run any node launch configuration
  - you should see `onDebugResolve:node` and `resolveDebugConfiguration:node` notifications
- test calling two different extension providers for debug type '*' supporting `onDebugResolve` :
  - install [on-debug-resolve-0.0.1.vsix](https://github.com/alvsan09/debug-activation-events/blob/master/on-debug-resolve-0.0.1.vsix)
  - install [on-debug-resolve-2.vsix](https://github.com/alvsan09/debug-activation-events/blob/master/on-debug-resolve-2.vsix)
  - try to run any node launch configuration
  - you should see `resolveDebugConfiguration(*)[0]` and `resolveDebugConfiguration(*)[1]` notifications, one for each of the installed `on-debug-resolve-*` extensions. ![config_providers_resolve](https://user-images.githubusercontent.com/76971376/159554236-f500838b-54e0-46f4-ae29-804768e0b667.png)
- test #9978, follow the instructions on the issue:
  - you should now see two terminals created one for the session and the other one triggered by the extension (there is still an issue here as the behavior does not match vscode (i.e. vscode does not mantain a session), however the original issue is resolved as we no longer create additional terminals i.e. one per registered debug adapter contributor).

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

colin-grant-work commented on packages/plugin-ext/src/main/browser/debug/debug-main.ts:
```diff
@@ -178,6 +181,27 @@ export class DebugMainImpl implements DebugMain, Disposable {
+            this.pluginDebugService.registerDebugConfigurationProvider(new PluginDebugConfigurationProvider(description, this.debugExt))
+        );
+
+        this.toDispose.push(Disposable.create(() => this.$unregisterDebugConfigurationProvider(handle)));
```
Perhaps `this.toDispose.push(toDispose)`? If it's disposed via an external call to `$unregisterDebugConfigurationProvider`, it will also be removed from `this.toDIspose`, so I believe the effect is the same.

colin-grant-work commented on packages/plugin-ext/src/main/browser/debug/plugin-debug-service.ts:
```diff
@@ -104,52 +99,72 @@ export class PluginDebugService implements DebugService, PluginDebugAdapterContr
+        for (const provider of pluginProviders) {
+            const result = await provider.provideDebugConfigurations(workspaceFolderUri);
+            if (result) {
+                results = results.concat(result);
```
Minor: I think `results.push(...result)` would be a little better than creating a new array each time.

colin-grant-work commented on packages/plugin-ext/src/main/browser/debug/plugin-debug-service.ts:
```diff
@@ -104,52 +99,72 @@ export class PluginDebugService implements DebugService, PluginDebugAdapterContr
-        return { type, configurations };
+        const configurationsMap = new Map<string, DebugConfiguration[]>();
+
+        await Promise.all(Array.from(pluginProviders, async provider => {
```
Minor: 
```suggestion
        await Promise.all(pluginProviders.map(async provider => {
```

colin-grant-work commented on packages/plugin-ext/src/plugin/node/debug/debug.ts:
```diff
@@ -331,15 +334,60 @@ export class DebugExtImpl implements DebugExt {
+        provider: theia.DebugConfigurationProvider
+        type: string
+    } {
+        const results = this.configurationProviders.filter(p => p.handle === handle);
+        if (results.length === 0) {
+            throw new Error('No Debug configuration provider found with given handle number: ' + handle);
+        }
+        const { provider, type } = results[0];
```
Minor:
```suggestion
        const record = this.configurationProviders.find(p => p.handle === handle);
        if (!record) {
            throw new Error('No Debug configuration provider found with given handle number: ' + handle);
        }
        const { provider, type } = record;
```

colin-grant-work reviewed (commented):
Regarding the testing steps, all but one worked for me, and the failing one may just be a misunderstanding of the conditions - could you elaborate a little on exactly when that message should appear?

- [x] onDebug appears for any event.
- [ ] onDebugInitialConfigurations appears when no launch.json + add config + markdown open. *This didn't appear to work for me*
- [x] onDebugResolve appears when running any configuration.
- [x] onDebugResolve:node appears when running a node configuration.
- [x] both onDebugResolve:* activation plugins are called.
- [x] multiple calls are not made to each provider for each resolution.

Regarding the code, it looks good for the most part. I had some comments about the location of some interfaces and a change to the `DebugService` interface, and some questions about some of the implementation details that mostly come from the old code.

alvsan09 reviewed (commented):
Addressing review comments, and reflected in latest commit!

alvsan09 commented:
Hi @colin-grant-work , 
I have addressed your comments and encapsulated my answers in a review to avoid sending individual notifications, 
I hope this is not confusing!
Thanks !

colin-grant-work commented on packages/debug/src/browser/debug-prefix-configuration.ts:
```diff
@@ -122,12 +122,11 @@ export class DebugPrefixConfiguration implements CommandContribution, CommandHan
-        const configurationsByType = await this.debugConfigurationManager.provideDynamicDebugConfigurations();
-        for (const typeConfigurations of configurationsByType) {
-            const dynamicConfigurations = typeConfigurations.configurations;
+        const map = await this.debugConfigurationManager.provideDynamicDebugConfigurations();
+        for (const [type, dynamicConfigurations] of map) {
```
It's true that the existing type doesn't ensure that all `DebugConfigurations` for a given heading are actually grouped together, and it's true that the only _actual_ implementation of the method currently lives in the browser, but I'd still prefer to avoid adding a type to the interface that absolutely _couldn't_ be implemented by the base `DebugService` on the backend. Given that, `Record<string, DebugConfiguration[] | undefined>` would achieve the goal of grouping configurations while retaining serializibility.

alvsan09 commented:
This PR has just been re-based to the latest master

paul-marechal commented on packages/plugin-ext/src/main/browser/debug/plugin-debug-service.ts:
```diff
@@ -104,52 +100,71 @@ export class PluginDebugService implements DebugService, PluginDebugAdapterContr
-            return contributor.provideDebugConfigurations && contributor.provideDebugConfigurations(workspaceFolderUri) || [];
-        } else {
+        const pluginProviders =
+            Array.from(this.configurationProviders.values()).
+                filter(
+                    p => p.triggerKind === DebugConfigurationProviderTriggerKind.Initial &&
+                    ( p.type === debugType || p.type === '*') &&
+                    !!p.provideDebugConfigurations);
```
The formatting looks weird here:

```suggestion
            Array.from(this.configurationProviders.values()).filter(p => (
                p.triggerKind === DebugConfigurationProviderTriggerKind.Initial &&
                (p.type === debugType || p.type === '*') &&
                p.provideDebugConfigurations
            ));
```

paul-marechal commented on packages/plugin-ext/src/main/browser/debug/plugin-debug-service.ts:
```diff
@@ -104,52 +100,71 @@ export class PluginDebugService implements DebugService, PluginDebugAdapterContr
             return this.delegated.provideDebugConfigurations(debugType, workspaceFolderUri);
         }
+
+        const results: DebugConfiguration[] = [];
+        Promise.all(pluginProviders.map(async p => {
+            const result = await p.provideDebugConfigurations(workspaceFolderUri);
+            if (result) {
+                results.push(...result);
+            }
+        }));
+
+        return results;
```
You forgot to `await Promise.all(...)`.

paul-marechal commented on packages/plugin-ext/src/main/browser/debug/plugin-debug-service.ts:
```diff
@@ -104,52 +100,71 @@ export class PluginDebugService implements DebugService, PluginDebugAdapterContr
-        const result: Promise<{ type: string, configurations: theia.DebugConfiguration[] }>[] = [];
+    async provideDynamicDebugConfigurations(): Promise<Record<string, DebugConfiguration[]>> {
+        const pluginProviders =
+            Array.from(this.configurationProviders.values()).
+                filter(
+                    p => p.triggerKind === DebugConfigurationProviderTriggerKind.Dynamic &&
+                    !!p.provideDebugConfigurations);
```
Formatting:

```suggestion
            Array.from(this.configurationProviders.values()).filter(p => (
                p.triggerKind === DebugConfigurationProviderTriggerKind.Dynamic &&
                p.provideDebugConfigurations
            ));
```

paul-marechal reviewed (commented):
Just did a high level pass.

alvsan09 commented:
> Just did a high level pass.
Updated in the latest commit!! Thanks !!

paul-marechal commented on packages/plugin-ext/src/plugin/node/debug/debug.ts:
```diff
@@ -331,15 +334,60 @@ export class DebugExtImpl implements DebugExt {
-                    result = result.concat(await provider.provideDebugConfigurations(this.toWorkspaceFolder(workspaceFolderUri)) || []);
-                }
+        const trigger = dynamic ? DebugConfigurationProviderTriggerKind.Dynamic : DebugConfigurationProviderTriggerKind.Initial;
+        const providers = this.configurationProviders.
+            filter(p =>
+                (p.type === debugType) &&
+                (p.trigger === trigger)).
+            map(p => p.provider);
```
In JavaScript the convention is to keep the `.method(` on the same line, unlike `.<\n>method`:

```suggestion
        const providers = this.configurationProviders
            .filter(p => (
                p.type === debugType &&
                p.trigger === trigger
            ))
            .map(p => p.provider);
```

paul-marechal commented on packages/plugin-ext/src/plugin/node/debug/debug.ts:
```diff
@@ -349,25 +397,22 @@ export class DebugExtImpl implements DebugExt {
-                        } catch (e) {
-                            console.error(e);
-                        }
+        const providers = this.configurationProviders.
+            filter(p =>
+                p.type === debugConfiguration.type || p.type === '*').
+            map(p => p.provider);
```
```suggestion
        const providers = this.configurationProviders
            .filter(p => p.type === debugConfiguration.type || p.type === '*')
            .map(p => p.provider);
```

alvsan09 commented:
@paul-marechal , 
The formatting remarks have been addressed in the latest commit !
Thanks !

colin-grant-work reviewed (approved):
All of the provided plugins work as expected. @paul-marechal, if you'd give it another pass, I'd appreciate it - it LGTM.

paul-marechal commented on packages/plugin-ext/src/main/browser/debug/plugin-debug-service.ts:
```diff
@@ -92,6 +78,16 @@ export class PluginDebugService implements DebugService, PluginDebugAdapterContr
+        return Disposable.create(() => this.unRegisterDebugConfigurationProvider(handle));
+    }
+
+    unRegisterDebugConfigurationProvider(handle: number): void {
```
It is more common to write `unregister` as a single word, rather than `unRegister`:

```suggestion
    unregisterDebugConfigurationProvider(handle: number): void {
```

paul-marechal commented on packages/plugin-ext/src/plugin/node/debug/debug.ts:
```diff
@@ -331,15 +334,58 @@ export class DebugExtImpl implements DebugExt {
+    async $resolveDebugConfigurationWithSubstitutedVariablesByHandle(
+        handle: number,
+        workspaceFolderUri: string | undefined,
+        debugConfiguration: theia.DebugConfiguration): Promise<theia.DebugConfiguration | undefined> {
```
Easier to read:

```suggestion
        debugConfiguration: theia.DebugConfiguration
    ): Promise<theia.DebugConfiguration | undefined> {
```

paul-marechal commented on packages/plugin-ext/src/plugin/node/debug/debug.ts:
```diff
@@ -331,15 +334,58 @@ export class DebugExtImpl implements DebugExt {
+    async $resolveDebugConfigurationByHandle(
+        handle: number,
+        workspaceFolderUri: string | undefined,
+        debugConfiguration: theia.DebugConfiguration): Promise<theia.DebugConfiguration | undefined> {
```
Easier to read:

```suggestion
        debugConfiguration: theia.DebugConfiguration
    ): Promise<theia.DebugConfiguration | undefined> {
```

paul-marechal commented on packages/plugin-ext/src/main/browser/debug/plugin-debug-configuration-provider.ts:
```diff
@@ -0,0 +1,67 @@
+    resolveDebugConfiguration: (folder: string | undefined, debugConfiguration: DebugConfiguration) => Promise<DebugConfiguration | undefined>;
+    resolveDebugConfigurationWithSubstitutedVariables: (folder: string | undefined, debugConfiguration: DebugConfiguration) => Promise<DebugConfiguration | undefined>;
+
+    constructor(description: DebugConfigurationProviderDescriptor,
+        protected readonly debugExt: DebugExt) {
```
Easier to read:

```suggestion
    constructor(
        description: DebugConfigurationProviderDescriptor,
        protected readonly debugExt: DebugExt
    ) {
```

paul-marechal commented on packages/plugin-ext/src/main/browser/debug/plugin-debug-configuration-provider.ts:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    get handle(): number {
+        return this._handle;
+    }
+
+    get type(): string {
+        return this._type;
+    }
+
+    get triggerKind(): DebugConfigurationProviderTriggerKind {
+        return this.trigger;
+    }
```
Since those fields are declared as `readonly` in `interface DebugConfigurationProvider` we can just convert them to `public` fields in this implementation. Code sites dealing with this class as a `DebugConfigurationProvider` won't be able to edit the fields because of the typings.

paul-marechal commented on packages/plugin-ext/src/main/browser/debug/plugin-debug-service.ts:
```diff
@@ -159,19 +174,21 @@ export class PluginDebugService implements DebugService, PluginDebugAdapterContr
-                    console.error(e);
+        const allProviders = Array.from(this.configurationProviders.values());
+        // Append debug type '*' at the end
+        const pluginProviders = allProviders.filter(p => p.type === config.type && !!p.resolveDebugConfigurationWithSubstitutedVariables);
+        pluginProviders.push(...allProviders.filter(p => p.type === '*' && !!p.resolveDebugConfigurationWithSubstitutedVariables));
```
Alternative writing, feel free to ignore:

```suggestion
        const providersForType = allProviders.filter(p => p.type === config.type && !!p.resolveDebugConfigurationWithSubstitutedVariables);
        const providersForAny = allProviders.filter(p => p.type === '*' && !!p.resolveDebugConfigurationWithSubstitutedVariables));
        const pluginProviders = providersForType.concat(providersForAny);
```

paul-marechal reviewed (commented):
Just a few remarks, but the logic looks good.

alvsan09 commented:
> Just a few remarks, but the logic looks good.
I have addressed all your comments except for the optional one which is an alternative I considered but I preferred the current one.
Thanks @paul-marechal  !!

paul-marechal reviewed (approved):
Code LGTM
